rust/worker/src/server.rs

```diff
@@ -182,32 +184,85 @@ impl WorkerServer {
             }
         };
 
-        let result = match self.system {
-            Some(ref system) => {
-                let orchestrator = HnswQueryOrchestrator::new(
-                    // TODO: Should not have to clone query vectors here
-                    system.clone(),
-                    query_vectors.clone(),
-                    request.k,
-                    request.allowed_ids,
-                    request.include_embeddings,
-                    segment_uuid,
-                    collection_uuid,
-                    self.log.clone(),
-                    self.sysdb.clone(),
-                    self.hnsw_index_provider.clone(),
-                    self.blockfile_provider.clone(),
-                    dispatcher,
-                    collection_version,
-                    log_position,
-                );
-                orchestrator.run().await
-            }
+        let system = match self.system {
+            Some(ref system) => system,
             None => {
+                tracing::error!("No system found");
                 return Err(Status::internal("No system found"));
             }
         };
 
+        let knn_filter_orchestrator = KnnFilterOrchestrator::new(
+            dispatcher.clone(),
+            // TODO: Load the configuration for this
+            1000,
+            FetchLogOperator {
+                log_client: self.log.clone(),
+                batch_size: 100,
+                skip: log_position as u32 + 1,
+                fetch: None,
+                collection: collection_uuid,
+            },
+            FetchSegmentOperator {
+                sysdb: self.sysdb.clone(),
+                hnsw: self.hnsw_index_provider.clone(),
+                blockfile: self.blockfile_provider.clone(),
+                knn: Some(IndexUuid(segment_uuid)),
+                metadata: None,
+                record: None,
+                collection: collection_uuid,
+                version: collection_version,
+            },
+            FilterOperator {
+                query_ids: Some(request.allowed_ids),
+                where_clause: None,
+            },
+        );
+
+        let knn_filter_output = match knn_filter_orchestrator.run(system.clone()).await {
+            Ok(output) => output,
+            Err(e) => {
+                tracing::error!("Error running orchestrator: {}", e);
+                return Err(Status::new(
+                    e.code().into(),
+                    format!("Error running orchestrator: {}", e),
+                ));
+            }
+        };
+
+        let embedding_dim = query_vectors[0].len();
```

Comment:
is it possible `query_vectors.len() == 0`?

Reply:
I'm unsure, but previously we are relying on this (line 230 before change)